### PR TITLE
test(server/runtime): make weak assertions bite across the runtime and services

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

Audit of all 72 test files under `src/server/runtime-handler`, `src/server/services`, `src/server/dev-server`, `src/server/context`, `src/server/project-env`, `src/server/shared` and `src/server/utils`.

Every change was **mutation checked**: the named source edit was applied, the test confirmed to fail, then reverted. 72 candidate findings, 70 confirmed after adversarial verification, 2 refuted.

## Recurring weaknesses fixed

**Misnamed cases that proved the opposite of their title.** Two RSC script handlers were named for a missing source and an unavailable esbuild, but exercised neither. They now spy on `readFile` and assert the compiled bundle is served verbatim without touching the filesystem.

**Tautologies and inequality assertions.** An adapter-factory test asserted `succeeded || threw`; it now pins `configOutcome` and the loaded config title, with a sibling case asserting the `cache-invariant-violation` slug. The browser-module cache replaced two inequalities with an exact stats literal, so failing to skip an oversized module now fails.

**Escaping never checked against a real payload.** Error pages now feed ``, `A & B <b>` and a closing `</script>` through both the styled page and the inline postMessage script, asserting the escaped forms are present and the raw markup absent anywhere in the document.

**Tracing asserted structurally.** `startRequestTracing` checked that a span object existed. It now installs a fake tracer and propagator, and asserts the W3C `traceparent` trace id is continued, the span is named `GET /test`, and a request without the header gets a different trace id.

**Branches no test reached:** the exact `CRITICAL` memory threshold boundary and the env threshold ladder, action roots that escape the project tree (with zero filesystem calls), a real-filesystem TypeScript middleware load leaving no temp files, `process.env` assigned outside any scope, and the hosted `waitForWarmup` gate on source-file collection.

## Source changes

Test seams only, no behaviour change:
- `renderer/memory/pressure.ts`: extracted the pure `classifyMemoryPressure` and made `parseEnvThreshold` take an injectable reader. The inlined integer parse is a faithful copy of `getEnvNumber`'s semantics (same `/^[+-]?\d+$/` guard, same `Number.isSafeInteger` check), so accepted values are unchanged.
- `server/utils/domain-lookup.ts`: `__injectFetchForTests`, matching the existing `__injectCacheForTests`.
- `channels/control-plane.ts`: exported the already-existing run operation route pattern.

## Verification

- Semantic unit-boundary audit: ok (base = merge-base with main), migration file untouched
- Sanitizer ratchet: ok 382/382 (down from 390; `ratchet.test.ts` passes), `test:layout` ok
- All eight repo ratchets and `docs:api-reference:check`: ok
- `deno fmt`, `deno lint`: clean
- **47/47** changed test files pass

### CODE REVIEW

File: src/channels/control-plane.ts [hunk 1]
```diff
@@ -36,7 +36,8 @@ export const DISPATCH_JWS_HEADER = "x-veryfront-dispatch-jws";
 /** The one route that accepts a signed channel dispatch envelope. */
 export const CHANNEL_INVOKE_PATH = "/channels/invoke";
 
-const CONTROL_PLANE_RUN_OPERATION_PATH =
+/** Matches the signed run operation routes (`execute`, `stream`, `resume`) of a control-plane run. */
+export const CONTROL_PLANE_RUN_OPERATION_PATH =
   /^\/api\/control-plane\/runs\/[^/]+\/(?:execute|stream|resume)$/u;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/channels/control-plane.ts]: Fixed in 43b9f879a. You are right that `./channels/control-plane` is a published entrypoint (deno.json line 106), so exporting the matcher widened the public API purely for a test.

Both route patterns now live in `src/channels/control-plane-routes.ts`, which is not listed in deno.json exports, and `control-plane.ts` imports them. The dispatch-ordering guard imports from the internal module, so it still reads the verb alternation and keeps its coverage.

Evidence: the public surface is unchanged, `deno task docs:api-reference:check` reports the reference current at 45 files, and `src/channels/control-plane.test.ts` plus `dispatch-exemption-ordering.test.ts` both pass.

File: src/server/shared/renderer/memory/pressure.test.ts [hunk 2]
```diff
@@ -1,10 +1,40 @@
...
 
 describe("server/shared/renderer/memory/pressure", () => {
+  afterEach(() => {
+    __injectDepsForTests(null);
+  });
+
   describe("shouldRejectDueToMemory", () => {
+    it("rejects at exactly the CRITICAL threshold and above", () => {
+      withHeapUsedPercent(DEFAULT_THRESHOLDS.CRITICAL);
+      assertEquals(shouldRejectDueToMemory(), true, "heap at CRITICAL (80) must reject");
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/shared/renderer/memory/pressure.test.ts]: Fixed in 43b9f879a. The boundary cases hard-coded 65/75/80 while the module resolves `THRESHOLDS` from `MEMORY_*_THRESHOLD` at load, so any configured value made the assertions wrong, exactly as you describe. The bare `79` and `64.9` and `99.5` literals had the same problem.

`THRESHOLDS` is now exported (internal; it does not appear in the API reference) and every boundary is expressed relative to the resolved values, with the resolved number interpolated into each message.

Evidence: `MEMORY_WARNING_THRESHOLD=30 MEMORY_HIGH_THRESHOLD=40 MEMORY_CRITICAL_THRESHOLD=50 deno task test:file src/server/shared/renderer/memory/pressure.test.ts` now passes; before this change those settings would have failed the CRITICAL boundary cases.

File: src/server/services/rendering/ssr.service.test.ts [hunk 3]
```diff
@@ -185,6 +185,23 @@ describe("server/services/rendering/ssr.service", () => {
...
+          true,
+          "heapLimitMB must come from getHeapStats, not a hardcoded zero",
+        );
+        assertEquals(status.heapUsedMB > 0, true, "heapUsedMB must come from getHeapStats");
+        assertEquals(
+          Math.abs(status.heapUsedPercent - (status.heapUsedMB / status.heapLimitMB) * 100) < 1,
+          true,
+          "heapUsedPercent must be the real ratio reported by getHeapStats",
+        );
+        assertEquals(status.shouldReject, false, "a healthy test process must not be shed");
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/server/services/rendering/ssr.service.test.ts]: Confirmed and fixed. Reproduced the exact case you named: with the old assertion in place, `MEMORY_CRITICAL_THRESHOLD=0 deno task test:file src/server/services/rendering/ssr.service.test.ts` failed on `reports the live heap statistics` (1 failed (3 steps)), because `parseEnvThreshold` accepts 0 and `shouldRejectDueToMemory()` then classifies every heap percentage as critical.

Rather than dropping the assertion, the shouldReject half now runs through the pressure module seam that already exists for this (`__injectDepsForTests` in src/server/shared/renderer/memory/pressure.ts) and asserts both sides of the resolved threshold: `THRESHOLDS.CRITICAL - 1` must not shed, `THRESHOLDS.CRITICAL` must shed. It is expressed against the resolved `THRESHOLDS.CRITICAL` rather than a hardcoded 80, so it holds under any configured value including 0. The live-statistics test keeps its heapUsedMB/heapLimitMB/heapUsedPercent assertions unchanged, since those do not depend on configuration.

Verified: the file passes both with the default threshold and under `MEMORY_CRITICAL_THRESHOLD=0`.

**Review by kwakayama — COMMENTED** [review]
No actionable findings.

Evidence: control-plane route matching remains method-gated in `src/channels/control-plane.ts:70`, with the private patterns extracted unchanged in `src/channels/control-plane-routes.ts:13`; the ordering inventory now guards every operation verb in `src/server/runtime-handler/dispatch-exemption-ordering.test.ts:230`. Memory-pressure thresholds and boundaries are explicitly exercised in `src/server/shared/renderer/memory/pressure.test.ts:75`, while domain lookup’s injected transport is reset and verifies cache normalization in `src/server/utils/domain-lookup.test.ts:89`. The sanitizer baseline reduction matches eight removed opt-outs.

Rubric: correctness 40/40, tests 20/20, reliability/security 15/15, maintainability 14/15, scope/docs 10/10.

Review-Gate:
Reviewer: Codex
Reviewed-SHA: 5d0643b12d409321e6fe7df9f8540c4892409942
Score: 99/100
Actionable-Findings: 0
Verdict: APPROVE

**Comment by kwakayama** [comment]
@codex review

File: src/server/utils/domain-lookup.test.ts [hunk 4]
```diff
@@ -80,6 +86,74 @@ describe("server/utils/domain-lookup", () => {
...
+
+    afterEach(() => {
+      __injectFetchForTests(null);
+      clearDomainCache();
+    });
+
+    it("returns null when the lookup API answers 404", async () => {
+      __injectFetchForTests(
+        (() => Promise.resolve(new Response(null, { status: 404 }))) as typeof fetch,
+      );
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/server/utils/domain-lookup.test.ts]: Fixed in 7890dac0e8. The domain lookup suite now scopes both requests through withMockFetch, which coordinates global fetch, the outbound host transport, and DNS resolution. The module-local mutable fetch seam was removed entirely. Evidence: the focused domain-lookup suite passes 23 steps, and repository search finds no remaining __injectFetchForTests references.

File: src/server/shared/renderer/memory/pressure.test.ts [hunk 5]
```diff
@@ -24,4 +70,72 @@ describe("server/shared/renderer/memory/pressure", () => {
...
+      assertEquals(
+        getMemoryPressureLevel(),
+        "normal",
+        `below WARNING (${DEFAULT_THRESHOLDS.WARNING}) is normal`,
+      );
+      withHeapUsedPercent(DEFAULT_THRESHOLDS.WARNING);
+      assertEquals(
+        getMemoryPressureLevel(),
+        "warning",
+        `exactly WARNING (${DEFAULT_THRESHOLDS.WARNING}) is warning`,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/server/shared/renderer/memory/pressure.test.ts]: Fixed in 7890dac0e8. The getMemoryPressureLevel ladder now injects explicit ordered thresholds (10/20/30) together with heap statistics, so independently configured and unordered MEMORY_* values cannot alter its expected levels. Evidence: the focused pressure suite passes 13 steps with MEMORY_WARNING_THRESHOLD=75, MEMORY_HIGH_THRESHOLD=75, and MEMORY_CRITICAL_THRESHOLD=75.

**Comment by kwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review

File: src/channels/control-plane.ts [hunk 6]
```diff
@@ -6,6 +6,10 @@ import { skillRegistry } from "#veryfront/skill/registry.ts";
 import { base64urlEncodeBytes } from "#veryfront/utils/base64url.ts";
 import { defineSchema, lazySchema } from "#veryfront/schemas/index.ts";
 import type { InferSchema, Schema } from "#veryfront/extensions/schema/index.ts";
+import {
+  CONTROL_PLANE_RUN_OPERATION_PATH,
+  CONTROL_PLANE_RUN_PATH,
+} from "./control-plane-routes.ts";
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/channels/control-plane.ts]: Fixed in `9486c05d45d31c9a5e1655c04c4936781d8c5136`: `src/channels/control-plane.ts` now imports the extracted route constants through `#veryfront/channels/control-plane-routes.ts`. Focused alias assertion, `deno fmt --check src/channels/control-plane.ts`, `deno lint src/channels/control-plane.ts`, and `deno task lint:module-boundaries` pass. Broader type/test execution remains blocked by pre-existing current-main React default-export diagnostics and the `node:util/types` runtime preflight, neither touched by this one-line import fix.

**Comment by kwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review

**Review by kwakayama — COMMENTED** [review]
## Codex scored review

Reviewed SHA: `63b24a5ec9b50d3229180ceee4b01b2706a53343`

No actionable findings. The fix moves only the two network-effectful domain-lookup cases into integration coverage, restoring the semantic unit-boundary contract while keeping pure unit cases colocated. Semantic audit (2,159 unit executables, 743 effect-bearing disposals), test-layout, 20-step unit coverage, 2-step integration coverage, formatting/lint, and diff checks pass; all review threads are resolved.

Score: 99/100
Verdict: APPROVE

Review-Gate:
Reviewer: Codex
Reviewed-SHA: 63b24a5ec9b50d3229180ceee4b01b2706a53343
Score: 99/100
Actionable-Findings: 0
Verdict: APPROVE

**Comment by kwakayama** [comment]
Reviewed the network-test relocation and corrected its cross-module import at 2c54e7549 to use the repository alias. Exact-head domain lookup unit (20 steps) and integration (2 steps) suites pass; fmt, lint, and typecheck are clean.

**Review by kwakayama — APPROVED** [review]
Reviewed exact head 2c54e7549. Network-bound domain lookup coverage is correctly integration-scoped, fetch mocking remains coordinated, and cross-module imports follow the repository boundary.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review
